lib/SABER/Z3ExprManager.cpp

```diff
@@ -0,0 +1,180 @@
+//===- LeakChecker.cpp -- Memory leak detector ------------------------------//
+//
+//                     SVF: Static Value-Flow Analysis
+//
+// Copyright (C) <2013-2017>  <Yulei Sui>
+//
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+//===----------------------------------------------------------------------===//
+
+/*
+ * Z3ExprManager.cpp
+ *
+ *  Created on: Jul 15, 2022
+ *      Author: Yulei, Xiao and Jiawei
+ */
+
+#include "Util/Z3ExprManager.h"
+#include "Util/Options.h"
+
+using namespace SVF;
+using namespace SVFUtil;
+
+Z3ExprManager *Z3ExprManager::z3ExprMgr = nullptr;
+u32_t Z3ExprManager::totalCondNum = 0;
+
+Z3ExprManager::Z3ExprManager() : sol(getContext()) {
+}
+
+Z3ExprManager::~Z3ExprManager() {
+}
+
+Z3ExprManager *Z3ExprManager::getZ3ExprMgr() {
+    if (z3ExprMgr == nullptr) {
+        z3ExprMgr = new Z3ExprManager();
+    }
+    return z3ExprMgr;
+}
+
+/// Create a fresh condition to encode each program branch
+Z3Expr Z3ExprManager::createFreshBranchCond(const Instruction *inst) {
+    u32_t condCountIdx = totalCondNum++;
+    Z3Expr expr = getContext().bool_const(("c" + std::to_string(condCountIdx)).c_str());
+    Z3Expr negCond = NEG(expr);
+    setCondInst(expr, inst);
+    setNegCondInst(negCond, inst);
+    z3ExprVec.push_back(expr);
+    z3ExprVec.push_back(negCond);
+    return expr;
+}
+
+// compute NEG
+Z3Expr Z3ExprManager::NEG(const Z3Expr &z3Expr) {
+    return !z3Expr;
+}
+
+// compute AND
+Z3Expr Z3ExprManager::AND(const Z3Expr &lhs, const Z3Expr &rhs) {
+    if (eq(lhs, getFalseCond()) || eq(rhs, getFalseCond())) {
+        return getFalseCond();
+    } else if (eq(lhs, getTrueCond())) {
+        return rhs;
+    } else if (eq(rhs, getTrueCond())) {
+        return lhs;
+    } else {
+        Z3Expr expr = lhs.getExpr() && rhs.getExpr();
+        // check subexpression size and option limit
+        if (getExprSize(expr) > Options::MaxZ3Size) {
```

Comment:
I think it is better to add a quick precondition `expr.getExpr().num_args() > Options::MaxZ3Size || getExprSize(expr) > Options::MaxZ3Size` since the getExprSize requires dfs-search